docs/secure-coding-guidelines.md

```diff
@@ -0,0 +1,99 @@
+      3. __Sensitive__ data (e.g. passwords, keys)
+   2. Data Security:
+      1. Encrypt __sensitive__ data prior to persisting
+      2. Limit the amount of time __sensitive__ data is decrypted
+      3. Do not bundle API tokens in client side applications unless you intend for them to be publicly accessible
+   3. Encryption:
+      1. __Sensitive__ data should be secured at rest and in-transit using standard encryption protocols (e.g. cryptography libraries, OS level key management systems, https, etc…)
+      2. Cryptography of __sensitive__ data should meet a minimum standard established by the security team during a threat assessment and annual reviews
+   4. Passwords:
+      1. Passwords used to decrypt content should adhere to security team recommendations during a threat assessment and annual reviews.
+
+3. Logging & Error Handling:
+
+   1. Logging:
+      1. Secrets and __sensitive__ data should not be logged
+      2. Console logging should be removed from production code using linters
+   2. Error Handling:
+      1. All errors should be handled
+      2. Errors should not leak data and be made more generic. This is especially important for code which handles data classified as __sensitive__.
+      3. Production versions of applications shall have logging disabled
+
+4. 3rd Party Integrations & Applications:
+   1. Authentication and Authorization:
+      1. Implement access controls and authentication mechanisms for accessing external data
+      2. Store and manage API keys securely when accessing external sources
+      3. API keys used in client applications shall have a reduced scope adhering to Principle of Least Privilege
+      4. If an API key is not intended to be publicly accessible, do not use it in a client application
+   2. Integration/Application Integrity:
+      1. Monitor integrations/applications for changes and security issues.
+   3. Permissions:
+      1. All applications/dapps/snaps shall adhere to the following to Principle of Least Privilege with a default state of allowing no permissions
+      2. Users shall provide consent for required permissions of applications/dapps/snaps
+
+5. Dependency Management:
+
+   1. Dependency Integrity:
+      1. Use Dependabot to keep all dependencies up to date and to automate checks for updates
+      2. Pin dependency versions or use a yarn.lock file
+   2. Avoid Deprecated and Unmaintained Packages:
+      1. Using Socket.dev to check the health and maintenance status of dependencies and seek alternatives when necessary
+   3. Review Dependency Licenses:
+      1. Ensure dependencies' licenses align with your project's requirements
+   4. Reduce Dependencies and Keep Them Minimal:
+      1. Minimize dependencies to reduce potential vulnerabilities
+   5. Audit and Monitor Dependencies:
+      1. Projects should use the npm audit, Dependabot and Socket.dev tooling for periodic automated auditing of dependencies
```

Comment:
```suggestion
      1. Projects should use tools (npm audit, Dependabot and Socket.dev) for periodic automated auditing of dependencies
      2. Use npm audit locally as a tool that informs you of the backlog of issues you have. Use dependabot security alerts to react to new vulnerability reports in your dependencies.
        - Triage dependabot PRs early, merge quickly if updates are minor. Don't hesitate to close them if not useful or could never be applied - but leave a comment why. Call attention to PRs where a security update is a major version update and cannot be easily introduced.
      3. If vulnerability management is particularly important for the project, run audit as a CI step and fail if it finds new issues. This may be prone to false positives and be blocked by transitive dependencies not being updated upstream. Use `resolutions` to override or `npm-audit-resolver` package to build a process around safely ignoring selected issues (requires discipline)
      4. Socket.dev will automatically report on suspicious dependencies in your pull requests. For best results use the following etiquette when dealing with it:
        - avoid merging the PR without addressing Socket warnings
        - avoid the ignore-all command unless you're introducing socket to a new project that's known to be safe (very rare)
        - always add a short comment explaining why you're ignoring a warning of potential malice/danger. eg. for a publisher change 'known maintainer' is enough of a comment indicating that you've verified the new publisher of the package is legitimate.
        - if Socket is being annoying, remember i can be configured. contact @naugtur (Zbyszek) and report if a warning category is useless.
```

Reply:
Thanks, great suggestions! I haven't taken this exactly as-is, but I've tried to capture the spirit of it here: 43e5282ad7cc2c6c1a78c66b68f8b95286b25056

It's been condensed a bit, and the Dependabot part is omitted. I omitted Dependabot because I tend not to think of it as a security tool; it's quite slow and unreliable, sometimes taking many days to create a PR to address vulnerabilities, and it will silently fail to address them sometimes. It's also often not configured or not configured correctly in our organization. I left a "Update dependencies quickly" bullet there to nudge people to follow your advice about responding quickly though.

Reply:
the dependabot part was based on actual problems we've recently had. dependabot is not a tool helping to react fast. It's a reminder and it does some of the work for you creating the PR. It's a tool helping not to stay many versions behind on stuff. And we've had dependabot PRs for security updates sit there for over a year.

Reply:
```
always add a short comment explaining why you're ignoring a warning of potential malice/danger. eg. for a publisher change 'known maintainer' is enough of a comment indicating that you've verified the new publisher of the package is legitimate.
```
became
```
If you've investigated a warning and found that it's not indicative of a malicious dependency, ignore it with a bot comment and explain your investigation with a short comment
```
The latter is IMHO too vague to be memorable and I'm afraid people won't internalize what it means and either ignore the recommendation or put too much effort into creating writeups leading to annoyance.

This is my main issue with this doc - it's too terse and open to interpretation and could use much more examples.

Reply:
Point taken on the wording, I'll restore something closer to yours. Can't recall the thinking behind that right now.

For the most part the terseness of the document isn't intended, we just haven't had time to expand on most points yet. More detail would definitely be appreciated.

Reply:
I appreciate that dependabot has been serving an important role absent other reminders about security problems, but I think there are good reasons to prefer alternatives. We haven't rolled out an npm audit CI job yet in libraries, but it's a far more reliable solution. Dependabot on the other hand is extremely slow, frequently fails to update things correctly, and will silently give up if it doesn't find a way to update something. It's not a reliable way to flag security advisories. I intended to add an npm audit CI job to the module template to serve that purpose.

Reply:
Dependabot is only useful together with the alerts, yes

Reply:
The example Socket.dev resolution was restored here, along with a second example: https://github.com/MetaMask/contributor-docs/pull/85

I started restoring the Dependabot advice in this PR but didn't get far: https://github.com/MetaMask/contributor-docs/pull/86

It felt wrong to include advice on triaging Dependabot issues given that it's not used in many repos at the moment, and even in those where it is enabled it's often not setup correctly (such that the Dependabot PRs typically don't pass CI checks).